### PR TITLE
Generate tool results when using structured result

### LINKED ISSUE #174: Agents generate a broken message history with structured responses
If an agent uses a structured result type, then the final message produced is a tool call, but without a corresponding tool result. This creates an issue if the message history is passed to another agent call. With OpenAI, it results in the following error: `"An assistant message with 'tool_calls' must be followed by tool messages responding to each 'tool_call_id'.`

If the result type is `str`, then this isn't an issue because no tool call is generated.

Here is an example that demonstrates this. A structured response is generated, and then its messages are used as the input history to a second response, which fails. 

```python
import pydantic_ai
from pydantic import BaseModel

class Demo(BaseModel):
    x: int
    
    
agent = pydantic_ai.Agent(model='openai:gpt-4o', result_type=Demo)

# first call
response = await agent.run('set x to any number between 1 and 100')

# second call, errors
await agent.run('set x to 5 more than the last value of x', message_history=response.new_messages())
```

### PR DESCRIPTION
**Pull request by jlowin**
Closes #174

### CODE REVIEW

File: pydantic_ai_slim/pydantic_ai/agent.py [hunk 1]
```diff
@@ -215,23 +215,27 @@ async def run(
...
 
                 with _logfire.span('handle model response', run_step=run_step) as handle_span:
-                    either = await self._handle_model_response(model_response, deps)
+                    responses = await self._handle_model_response(model_response, deps)
 
-                    if isinstance(either, _MarkFinalResult):
-                        # we have a final result, end the conversation
-                        result_data = either.data
+                    # Add all messages to the conversation
+                    messages.extend(r for r in responses if isinstance(r, _messages.Message))
```
Comment (samuelcolvin) [inline pydantic_ai_slim/pydantic_ai/agent.py]: we extract the `Message`s from the list 3 times, maybe better to return `tuple[_MarkFinalResult[ResultData] | None, list[_messages.Message]]`?

Then we don't have to do any extraction aftwards.

**Review by samuelcolvin — COMMENTED** [review]
Overall LGTM, some tests are failing, but you can probably fix most of them by running `uv run pytest --fix`.

LMK if you want me to finish this off.

**Comment by samuelcolvin** [comment]
We should also test if the responses need to be adjacent in the messages, I didn't think they needed to be, but @sudoskys suggested they do. That can be a separate PR.

**Comment by jlowin** [comment]
Looking at those tests now and I like your suggestion about the tuple. 

Responses do need to be adjacent for OpenAI, not sure if others have the same requirement (message expectations vary). 

The OpenAI requirement is: a message with one or more tool calls must be followed immediately and only by messages containing responses to *every tool call*. For example if 2 parallel tool calls are made, the next two messages must be responses corresponding to those two tool call ids.

**Comment by jlowin** [comment]
I think this is GTG. In testing I identified another situation that can result in the same error message, in which the model requests a "regular" tool call and a "final" tool call in parallel in the same payload. Even though this PR adds the `ToolResult` for the "final" tool, the regular tool never gets a result and subsequent calls will fail. I will try to address in a separate PR.

**Comment by samuelcolvin** [comment]
> Even though this PR adds the

I think that's what's reported in #161. There's some more discussion needed in that case.

**Comment by jlowin** [comment]
Ok cool, I have a proposal I'll add on that issue.

**Comment by samuelcolvin** [comment]
Thanks so much @jlowin.

**Pull request merged**